gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala

```diff
@@ -0,0 +1,140 @@
+      } else false
+    } else {
+      // The transformable tag is already set before
+      TransformHints.isTransformable(agg)
+    }
+  }
+
+  private def isPartialAgg(partialAgg: BaseAggregateExec, finalAgg: BaseAggregateExec): Boolean = {
+    // TODO: now it can not support to merge agg which there are the filters in the aggregate exprs.
+    if (
+      partialAgg.aggregateExpressions.forall(x => x.mode == Partial && x.filter.isEmpty) &&
+      finalAgg.aggregateExpressions.forall(x => x.mode == Final && x.filter.isEmpty)
+    ) {
+      (finalAgg.logicalLink, partialAgg.logicalLink) match {
+        case (Some(agg1), Some(agg2)) => agg1.sameResult(agg2)
+        case _ => false
+      }
+    } else {
+      false
+    }
+  }
+
+  override def apply(plan: SparkPlan): SparkPlan = PhysicalPlanSelector.maybe(session, plan) {
+    if (!BackendsApiManager.getSettings.mergeTwoPhasesHashBaseAggregateIfNeed()) {
+      plan
+    } else {
+      plan.transformDown {
+        case hashAgg @ HashAggregateExec(
+              _,
+              isStreaming,
+              _,
+              _,
+              aggregateExpressions,
+              aggregateAttributes,
+              _,
+              resultExpressions,
+              child: HashAggregateExec)
+            if !isStreaming && isTransformable(hashAgg) && isTransformable(child) =>
+          if (isPartialAgg(child, hashAgg)) {
+            // convert to complete mode aggregate expressions
+            val completeAggregateExpressions = aggregateExpressions.map(_.copy(mode = Complete))
+            hashAgg.copy(
+              requiredChildDistributionExpressions = None,
+              groupingExpressions = child.groupingExpressions,
+              aggregateExpressions = completeAggregateExpressions,
+              initialInputBufferOffset = 0,
+              child = child.child
+            )
+          } else {
+            hashAgg
+          }
+        case objectHashAgg @ ObjectHashAggregateExec(
+              _,
+              isStreaming,
+              _,
+              _,
+              aggregateExpressions,
+              aggregateAttributes,
+              _,
+              resultExpressions,
+              child: ObjectHashAggregateExec)
+            if !isStreaming && isTransformable(objectHashAgg) && isTransformable(child) =>
+          if (isPartialAgg(child, objectHashAgg)) {
+            // convert to complete mode aggregate expressions
+            val completeAggregateExpressions = aggregateExpressions.map(_.copy(mode = Complete))
+            objectHashAgg.copy(
+              requiredChildDistributionExpressions = None,
+              groupingExpressions = child.groupingExpressions,
+              aggregateExpressions = completeAggregateExpressions,
+              initialInputBufferOffset = 0,
+              child = child.child
+            )
+          } else {
+            objectHashAgg
+          }
+        case plan: SparkPlan => plan
```

Comment:
can we also handle SortAggregate ? it is possible that there is no shuffle and sort between two SortAggregate.

Reply:
added SortAggregate